core/src/modules/cls.ts

```diff
@@ -31,23 +32,94 @@ export namespace CLS {
    * Wrap an Async function f in such a way that all enqueued afterCommit / enqueueTasks during invocation will be run afterwords
    * Returns the return value of function f
    */
-  export async function wrap(f: Function, catchError = false) {
+  export interface CLSWrapMethod {
+    (
+      f: Function,
+      options?: {
+        catchError?: boolean;
+        lock?: boolean;
+        write?: boolean;
+        priority?: boolean;
+      }
+    ): Promise<any>;
+  }
+  export const wrap: CLSWrapMethod = async (f, options = {}) => {
+    const { catchError, write, priority } = options;
+    let { lock } = options;
+    const dialect = api.sequelize.options.dialect;
+    if (!lock && dialect === "sqlite" && write) {
+      lock = true;
+    }
+    let releaseLock = null;
     try {
-      let runResponse: any;
-      let afterCommitJobs: Array<Function> = [];
-
-      await api.sequelize.transaction(async (t: Transaction) => {
-        runResponse = await f(t);
-        afterCommitJobs = getNamespace().get("afterCommitJobs");
-      });
-
-      for (const i in afterCommitJobs) await afterCommitJobs[i]();
-
+      if (lock) {
+        releaseLock = await getLock(priority);
+      }
+      const runResponse = await wrapInternal(f, options);
       return runResponse;
     } catch (error) {
       if (catchError) return error;
       throw error;
+    } finally {
+      if (releaseLock) {
+        await releaseLock();
+      }
+    }
+  };
+
+  const wrapInternal: CLSWrapMethod = async (f, options = {}) => {
+    const { write } = options;
+    let runResponse: any;
+    let afterCommitJobs: Array<Function> = [];
+
+    const transOptions: any = {};
+    const dialect = api.sequelize.options.dialect;
+    if (dialect === "sqlite" && write) {
+      transOptions.type = "IMMEDIATE";
+    }
+    await api.sequelize.transaction(transOptions, async (t: Transaction) => {
+      runResponse = await f(t);
+      afterCommitJobs = getNamespace().get("afterCommitJobs");
+    });
+
+    for (const i in afterCommitJobs) {
+      await afterCommitJobs[i]();
+    }
+
+    return runResponse;
+  };
+
+  async function getLock(priority) {
+    // check more often in the ui to get the lock
+    const longEnough = 30 * 1000; // 30 seconds
+    const sleepTime = priority ? 1 : 100; // check more often if it's a priority
+    const ttl = longEnough + 1;
+    const maxAttempts = 300; // hard-coded in the locks.ts code
+    // might have to check more times, so start the count negative to get to 30 seconds
+    const attempts = 0 - longEnough / sleepTime + maxAttempts;
```

Comment:
less then ideal API for `waitForLock` but out of scope for this PR, so made this affordance.